src/behavior_oracle.cpp

```diff
@@ -0,0 +1,138 @@
+#include "behavior_oracle.h"
+
+#include "behavior.h"
+#include "bodypart.h"
+#include "itype.h"
+#include "player.h"
+#include "weather.h"
+
+#include <functional>
+
+using namespace behavior;
+
+status_t oracle_t::return_running() const
+{
+    return running;
+}
+
+// To avoid a local minima when the character has access to warmth in a shelter but gets cold
+// when they go outside, this method needs to only alert when travel time to known shelter
+// approaches time to freeze.
+status_t character_oracle_t::needs_warmth_badly() const
+{
+    const player *p = dynamic_cast<const player *>( subject );
+    // Use player::temp_conv to predict whether the Character is "in trouble".
+    for( const body_part bp : all_body_parts ) {
+        if( p->temp_conv[ bp ] <= BODYTEMP_VERY_COLD ) {
+            return running;
+        }
+    }
+    return success;
+}
+
+status_t character_oracle_t::needs_water_badly() const
+{
+    // Check thirst threshold.
+    if( subject->get_thirst() > 520 ) {
+        return running;
+    }
+    return success;
+}
+
+status_t character_oracle_t::needs_food_badly() const
+{
+    // Check hunger threshold.
+    if( subject->get_hunger() >= 300 && subject->get_starvation() > 2500 ) {
+        return running;
+    }
+    return success;
+}
+
+status_t character_oracle_t::can_wear_warmer_clothes() const
+{
+    const player *p = dynamic_cast<const player *>( subject );
+    // Check inventory for wearable warmer clothes, greedily.
+    // Don't consider swapping clothes yet, just evaluate adding clothes.
+    for( const auto &i : subject->inv.const_slice() ) {
+        const item &candidate = i->front();
+        if( candidate.get_warmth() > 0 || p->can_wear( candidate ).success() ) {
+            return running;
+        }
+    }
+    return failure;
+}
+
+status_t character_oracle_t::can_make_fire() const
+{
+    // Check inventory for firemaking tools and fuel
+    bool tool = false;
+    bool fuel = false;
+    for( const auto &i : subject->inv.const_slice() ) {
+        const item &candidate = i->front();
+        if( candidate.has_flag( "FIRESTARTER" ) ) {
+            tool = true;
+            if( fuel ) {
+                return running;
+            }
+        } else if( candidate.flammable() ) {
+            fuel = true;
+            if( tool ) {
+                return running;
+            }
+        }
+    }
+    return success;
```

Comment:
We get there if we don't have FIRESTARTER or FUEL. Shouldn't it return failure then?